### PR TITLE
going around a (maybe slick's bug) that this query will always return a value…

### PR DESCRIPTION
**Pull request by eishay**
…, may be null. Doing a firstOption may return a Some(null) value

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 1]
```diff
@@ -524,7 +524,8 @@ class KeepRepoImpl @Inject() (
 
   def getDateLastManualKeep(userId: Id[User])(implicit session: RSession): Option[DateTime] = {
     import com.keepit.common.db.slick.StaticQueryFixed.interpolation
-    sql"""select max(kept_at) from bookmark where user_id = $userId and source in ('keeper', 'mobile', 'email', 'site')""".as[DateTime].firstOption
+    //going around a (maybe slick's bug?) that this query will always return a value, may be null. Doing a firstOption may return a Some(null) value
+    Option(sql"""select max(kept_at) from bookmark where user_id = $userId and source in ('keeper', 'mobile', 'email', 'site')""".as[DateTime].first)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:528]: Can't we do this directly in Slick?

`val sources: Set[KeepSource] = Set(KeepSource.keeper)`
`(for (b <- rows if b.userId === userId && b.source.inSet(sources)) yield b.keptAt).max.run`

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 2]
```diff
@@ -524,7 +524,8 @@ class KeepRepoImpl @Inject() (
 
   def getDateLastManualKeep(userId: Id[User])(implicit session: RSession): Option[DateTime] = {
     import com.keepit.common.db.slick.StaticQueryFixed.interpolation
-    sql"""select max(kept_at) from bookmark where user_id = $userId and source in ('keeper', 'mobile', 'email', 'site')""".as[DateTime].firstOption
+    //going around a (maybe slick's bug?) that this query will always return a value, may be null. Doing a firstOption may return a Some(null) value
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:527]: Technically, this is correct behavior. You're asking for a `DateTime`, and saying "if there is a row returned, give me `Some(the first one)`; else, give me `None`". Compare these two queries:

```
select max(kept_at) from bookmark where user_id=2; -- returns a row, with NULL
select kept_at from bookmark where user_id=2; -- returns no row
```

Serialize as `Option[DateTime]` to get what you're expecting.

**Comment by eishay** [comment]
I guess we could, didn't make an effort since it seems much more readable in plain sql. makes sense to change if it won't return Some(null)

**Comment by ryanpbrewster** [comment]
If you wanted to be even more Scala-like, I think that
`rows.filter(k => k.userId === userId && k.source.inSet(sources)).map(_.keptAt).max.run` (or something like that) ought to work.

**Comment by eishay** [comment]
fancy :-)

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala (hunk 2 above)
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:527]: cool, thanks for clarifying.

**Pull request merged**